Enable C++20

rleh opened the pull request:
- [x] Update ARM GCC to version 10
- [x] Use new Docker images (hosted on Github Container Registry ghcr.io)
- [x] Update installation guide
- [x] Enable C++20

rleh committed:
[ci] Use new docker images from ghcr.io instead of Dockerhub

rleh committed:
[docs] Update installation guide for ARM GCC 10

rleh committed:
[tools][doc] Enable C++20

rleh committed:
[ci] Update TravisCI and GH Action on Windows to ARM GCC 10

salkinium commented on docs/src/guide/installation.md:
```diff
@@ -53,14 +53,14 @@ For ARM Cortex-M targets:
+		export PATH="/opt/gcc-arm-none-eabi-10-2020-q4-major/bin:$PATH"
 
 For Atmel/Microchip AVR targets:
 
```
URL changed below

rleh replied:
Ahhh, got it...

salkinium reviewed (changes requested):
MacOS CI may fail until homebrew is updated.

rleh commented:
```
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(event_groups.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(list.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(port.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(queue.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(stream_buffer.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(tasks.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: build/release/modm/libmodm.a(timers.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: multiple definition of `modm::platform::fcpu'; build/release/modm/libmodm.a(croutine.o):/root/project/examples/nucleo_f103rb/rtos/modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:39: first defined here
collect2: error: ld returned 1 exit status
scons: *** [build/release/rtos.elf] Error 1
```

salkinium commented:
Oh yeah, I forgot C++20 deprecated `volatile`. Not entirely sure if we can fix this at all (considering the volatile comes from extern "C" headers) so we have to just ignore the warning?

rleh committed:
[tools] Disable volatile warning

rleh commented:
### Errors
* Cortex-M on WIndows: Zip folder structure changed. Fixed!
* Linux running on Aarch64 (Travis): GCC10 (hosted) has to be installed. Fixed!
* MacOS: Waiting for osx-cross/homebrew-arm#16...
* Cortex-M on Linux: see comment above
* Windows: WTF?!?
```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "../tools/scripts/examples_compile.py", line 113, in <module>
    exit(compile_examples(sys.argv[1:]))
  File "../tools/scripts/examples_compile.py", line 99, in compile_examples
    projects = pool.map(build, projects)
  File "C:\hostedtoolcache\windows\Python\3.8.6\x64\lib\multiprocessing\pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "C:\hostedtoolcache\windows\Python\3.8.6\x64\lib\multiprocessing\pool.py", line 771, in get
    raise self._value
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb7 in position 118: invalid start byte
Error: Process completed with exit code 1.
```

rleh commented:
### STM32 Unittests
FLASH' overflowed by 10424 bytes

```
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: release/test.elf section `.text' will not fit in region `FLASH'
/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: region `FLASH' overflowed by 10424 bytes
```

salkinium commented:
> FLASH' overflowed by 10424 bytes

Yes, the F103 was like 98% full before… so I guess there's been a little regression in GCC code size?

rleh commented:
There is already a lot disabled for STM32F1: https://github.com/modm-io/modm/blob/develop/test/config/nucleo-f103.xml
```xml
  <modules>
    <module>modm:platform:heap</module>
    <!-- Not everything fits -->
    <!-- <module>modm-test:test:architecture</module>
    <module>modm-test:test:communication</module>
    <module>modm-test:test:container</module>
    <module>modm-test:test:driver</module>
    <module>modm-test:test:io</module> -->

    <module>modm-test:test:platform:**</module>
    <module>modm-test:test:processing</module>
    <module>modm-test:test:ui</module>
    <module>modm-test:test:math</module>
  </modules>
```

chris-durand commented:
> Oh yeah, I forgot C++20 deprecated `volatile`. Not entirely sure if we can fix this at all (considering the volatile comes from extern "C" headers) so we have to just ignore the warning?

Only some parts are deprecated of course. Otherwise the compiler would optimize away all hardware register accesses. There is still no other way to tell the compiler not to mess with your memory mapped IO. They deprecated all the insane nonsensical parts no-one should ever use, like volatile qualified member functions, volatile return values, volatile function parameters, etc.

The annoying part is they also deprecated some parts which are ubiquitously used in vendor provided code. Compound operations on volatiles like `SOME_REGISTER |= (1 << 4)` are deprecated. There are a lot of embedded people who consider that part a mistake.

chris-durand commented:
From P1152R4 "Deprecating volatile"
```
We propose deprecating most of volatile. See § 3 Wording for the details.

The proposed deprecation preserves the useful parts of volatile, and removes the dubious / already broken ones.
```

chris-durand commented:
Reddit flamewar discussion on that topic: [Compound assignment to volatile must be un-deprecated
](https://www.reddit.com/r/cpp/comments/jswz3z/compound_assignment_to_volatile_must_be/)

salkinium commented:
> Summary of the comments:
> - "I write userspace applications with the standard library and know about std::atomic<>, you know about std::atomic, right?"
> - "You, an embedded developer, don't know how stores/loads interact with interrupts - you clearly don't know how invalid your code is"
> - "Why not write wrapper functions for the entire chip API, in your heavily-constrained embedded development environment. You're using gcc/llvm, right? It'll probably all be inlined and elilded, probably."
> - "Here's a bit-twiddling refactor of your version. I did this myself. It doesn't address the complaint, but it does save a load. Not enough loads to actually avoid a race with an interrupt, but it does save one."
> - "The chip manufacturer should just make their C headers C++20 compatible - unless the chip manufacturer is anything like 90 % of chip manufacturers, which typically only explicitly support an approved dialect of C"
> - "God, these embedded manufacturers should just get with the times and write modern C++. These potentially breaking changes are great for C++" ... "but don't move to an entirely different language like Rust because you'd be losing out on the backwards-compat benefits you want from C++"
> - "The C++ standards committee should implement a new language feature that enables C++ programmers to do stuff that's possible in C89. Also, they should deprecate bitfields, I don't like bitfields."

Scathing…

rleh committed:
[ci] Windows: GCC 10 from winlibs and PYTHONIOENCODING=utf-8

chris-durand commented:
We might have a few `[[nodiscard]]` in the wrong place:
```
In file included from modm/src/modm/platform.hpp:15,
                 from modm/src/modm/board/board.hpp:17,
                 from modm/src/modm/board.hpp:12,
                 from main.cpp:15:
modm/src/modm/platform/can/can_1.hpp:111:49: warning: 'nodiscard' attribute can only be applied to functions or to class or enumeration types [-Wattributes]
  111 |     bool overwriteOnOverrun = true) [[nodiscard]]
```
It has to go to the front. I am quite sure written like above it does nothing despite giving the warning about misuse of attributes.

chris-durand commented:
I'll have a look at the `[[nodiscard]]`s

rleh commented:
Now another error on Windows hosted:

```
 modm\src\modm\platform\core\assert.cpp: In function 'void modm_abandon(const modm::AssertionInfo&)':
modm\src\modm\platform\core\assert.cpp:60:61: error: unknown conversion type character 'l' in format [-Werror=format=]
   60 |  if (info.context != uintptr_t(-1)) { MODM_LOG_ERROR.printf(" @ %p (%" PRIuPTR ")", (void *)info.context, info.context); }
      |                                                             ^~~~~~~~~~~~~~~~~~~~~~
modm\src\modm\platform\core\assert.cpp:60:61: error: too many arguments for format [-Werror=format-extra-args]
cc1plus.exe: some warnings being treated as errors
```

Has anyone seen `character 'l'`?

rleh commented:
Probably because I exchanged MinGW32 with a gcc-10 based Mingw64...

chris-durand commented:
> Has anyone seen `character 'l'`?

Don't you see the `l` in `PRIuPTR` :wink: ?

 `PRIuPTR` is for `uintptr_t`, probably `lu`, format for `unsigned long`.

rleh commented:
https://stackoverflow.com/a/55099671

> Use %I64u on Windows, or just use inttypes.h PRIuMAX.

Or: https://github.com/dosbox-staging/dosbox-staging/issues/64#issuecomment-560158756

rleh commented:
:unamused: 
```
 modm\src/modm/platform/can/can_bit_timings.hpp:80:22: error: call to non-'constexpr' function 'double fabs(double)'
   80 |    float error = fabs(1 - intPrescaler/idealPrescaler);
...
c:\winlibs-gcc\mingw64\x86_64-w64-mingw32\include\math.h:220:31: note: 'double fabs(double)' declared here
  220 |   __CRT_INLINE double __cdecl fabs (double x)
```

chris-durand commented:
fabs is not required to be constexpr, not even in C++20 but in libstdc++ it actually is. Should I fix it by putting a constexpr version in there?
Like:
```
template <typename Float>
    requires std::is_floating_point_v<Float>
constexpr Float constexpr_fabs(Float number)
{
    if (number >= 0) {
        return number;
    } else 
        return -number;
    }
}
```

rleh commented:
Sehr gerne!

chris-durand commented:
Ok, I'll do it.

chris-durand committed:
[platform] Fix use of fabs in constexpr context

chris-durand commented:
Could you have a look at it and cherry-pick the commit? https://github.com/chris-durand/modm/commit/6578b2461b52132516dc4c99b381aa3328a7c808

rleh committed:
[ci] MacOS: Use our own homebrew-arm fork

rleh commented:
MacOS still ~is~ has a problem:
`avr-g++: error: unrecognized command line option '-std=c++20'; did you mean '-std=c++2a'?`

rleh committed:
[ci] MacOS: Use our own homebrew-avr fork

rleh commented:
Looks like avr-gcc 10.2.0 on MacOS is not that easy (https://github.com/osx-cross/homebrew-avr/pull/194).

I would suggest to drop AVR support on MacOS until someone provides an updated toolchain for MacOS.
What do you think @salkinium?

salkinium commented:
I'll take a look at avr-gcc v10. I'd like to not have a blocker like that in modm.

salkinium reviewed (approved):
Thanks, and I'll check what the issue with avr-gcc is.

rleh commented:
I started working on avr-gcc 10 for MacOS: modm-ext/homebrew-avr#2.

salkinium committed:
[tools] Use avr-objdump for size info

salkinium commented:
I changed the size tool info, [since `avr-size` now shows Device Unknown, since afaik the size and device table has been moved into avr-gcc instead of avr-binutils](https://www.avrfreaks.net/comment/2446681#comment-2446681), and can be retrieved via `avr-objdump -Pmem-usage file.elf` without also providing the device name.

chris-durand commented:
We will also need to update the AVR libstdc++ to the C++20 version. I'll have a look at it tomorrow.

rleh committed:
[ci] MacOS: Temporarily disable AVR testing because toolchain is not available

rleh committed:
[ext] Update avr-libstdcpp

rleh committed:
[ci] Windows: Fix Winlibs toolchain download link

rleh committed:
[platform] avr: I²C fix missing include <algorithm>

rleh committed:
[test] Fix libstdc++ test by removing <clocale> header

rleh committed:
[test] avr: split arduino-nano unittest into 8 parts

rleh commented:
Some unit-test fail on AVR (Arduino Nano):
```
Unittests (Jan  5 2021, 17:46:48)
>>> atomic_queue_test
>>> can_message_test
>>> clock_test
>>> flash_test
>>> i2c_test
>>> interface_test
>>> register_test
>>> slave_test
FAIL: slave_test:118 : 0 == 1
FAIL: slave_test:140 : 0 == 2
FAIL: slave_test:148 : true == false
FAIL: slave_test:151 : 1 == 2
FAIL: slave_test:154 : 
[1, 163]
[205, 171]
FAIL: slave_test:166 : 0 == 3
FAIL: slave_test:190 : 0 == 4
FAIL: slave_test:191 : 0 == 39030
FAIL: slave_test:199 : true == false
FAIL: slave_test:202 : 1 == 4
FAIL: slave_test:204 : 
[1, 9, 0, 222]
[120, 86, 52, 18]

Failed 11 of 216 tests
FAIL!

Unittests (Jan  5 2021, 17:38:45)
>>> bounded_deque_test
>>> bouUnittests (Jan  5 2021, 17:38:45)
>>> bounded_deque_test
>>> bounded_queue_test
>>> bounded_stack_test
>>> doubly_linked_list_test
>>> dynamic_array_test
>>> linked_list_test
>>> pair_test

Passed 486 tests
OK!

Unittests (Jan  5 2021, 17:50:44)
>>> ad7280a_test
>>> block_allUnittests (Jan  5 2021, 17:50:44)
>>> ad7280a_test
>>> block_allocator_test
FAIL: block_allocator_test:97 : 1 == 0
FAIL: block_allocator_test:98 : 1 == 0
FAIL: block_allocator_test:97 : 2 == 0
FAIL: block_allocator_test:98 : 2 == 0
FAIL: block_allocator_test:97 : 3 == 0
FAIL: block_allocator_test:98 : 3 == 0
FAIL: block_allocator_test:97 : 1 == 0
FAIL: block_allocator_test:98 : 1 == 0
FAIL: block_allocator_test:97 : 2 == 0
FAIL: block_allocator_test:98 : 2 == 0
>>> can_lawicel_formatter_test
>>> drv832x_spi_test
>>> ltc2984_test
>>> mcp2515_can_bit_timings_test
>>> spi_device_test

Failed 10 of 296 tests
FAIL!

Unittests (Jan  5 2021, 17:51:13)
>>> can_bit_timings_test
>>> iUnittests (Jan  5 2021, 17:51:13)
>>> can_bit_timings_test
>>> io_stream_test

Passed 180 tests
OK!

Unittests (Jan  5 2021, 17:38:45)
>>> button_group_test
>>> buttUnittests (Jan  5 2021, 17:38:45)
>>> button_group_test
>>> button_test
>>> time_test
FAIL: time_test:33 : 984534400 == 1000000000
FAIL: time_test:52 : 1073676288 == 1073741824
FAIL: time_test:71 : 1300246921 == 1311738121
FAIL: time_test:89 : 1329354884 == 1333329284
FAIL: time_test:108 : 3130101453 == 3141592653
FAIL: time_test:123 : 12 == 9
FAIL: time_test:124 : 5 == 8
FAIL: time_test:140 : 2 == 10
FAIL: time_test:141 : 1 == 0
FAIL: time_test:157 : 92 == 27
FAIL: time_test:158 : 12 == 6
FAIL: time_test:174 : 8 == 2
FAIL: time_test:175 : 8 == 3
FAIL: time_test:192 : 86 == 21
FAIL: time_test:193 : 12 == 6

Failed 15 of 54 tests
FAIL!
```

Some other tests do not fit on my Arduino Nano: `modm-test:test:communication:xpcc`, `<module>modm-test:test:processing</module>` and `<module>modm-test:test:math</module>`.

cc @chris-durand

chris-durand commented:
https://github.com/modm-io/modm/pull/528 fixes timestamp_test in processing. With the fix processing passes.

rleh commented:
> _>>>_ block_allocator_test
> FAIL: block_allocator_test:97 : 1 == 0
> FAIL: block_allocator_test:98 : 1 == 0

is caused by

```c++
void
BlockAllocatorTest::testAlignment()
{
	uint8_t *heap = new uint8_t[512];

	for (uint_fast8_t misalignment = 0; misalignment < 6; ++misalignment)
	{
		modm::BlockAllocator<uint16_t, 8> allocator;
		allocator.initialize(heap + misalignment, heap + 512);

		TEST_ASSERT_EQUALS(allocator.getAvailableSize(), 496U);

		void* firstBlock = allocator.allocate(12);
		void* secondBlock = allocator.allocate(12);

		TEST_ASSERT_EQUALS( ((uintptr_t)firstBlock) % 4, 0U);
		TEST_ASSERT_EQUALS( ((uintptr_t)secondBlock) % 4, 0U);
	}

	delete[] heap;
}
```

Mhmmm...?
Maybe memory exhaustion on my _atmega328p_?

chris-durand commented:
The block allocator test also fails on my ATmega2560. If it was memory exhaustion, you would end up in an assertion handler when you allocate the memory with new there: `uint8_t *heap = new uint8_t[512];`

rleh commented:
Unittests on Nucleo-F411 also do not pass:
```
Unittests (Jan  5 2021, 18:20:59)
[...]

>>> LUDecomposition_test
>>> ad7280a_test
>>> angle_test
FAIL: angle_test:40 : -3.14159e+00 == 3.14159e+00
>>> arithmetic_traits_test
>>> atomic_queue_test
>>> bit_operation_test
>>> block_allocator_test
>>> bme280_test
>>> bmp085_test
>>> bounded_deque_test
>>> bounded_queue_test
>>> bounded_stack_test
>>> button_group_test
>>> button_test
>>> can_bit_timings_test
>>> can_connector_base_test
>>> can_connector_test
>>> can_lawicel_formatter_test
>>> can_message_test
>>> circle_2d_test
>>> clock_test
>>> debounce_test
>>> dispatcher_test
>>> doubly_linked_list_test
>>> drv832x_spi_test
>>> dynamic_array_test
>>> endianness_test
>>> fir_test
>>> flash_test
>>> gpio_sampler_test
Channel 0:
-1658617419
1658617554
-1658713636
1658809707
-1658905816
Channel 1:
1658617419
-1659001896
1659097977
-1659194049
1659290121
>>> i2c_test
>>> interface_test
>>> interface_test
>>> io_stream_test
>>> lagrange_interpolation_test
>>> line_2d_test
>>> line_segment_2d_test
>>> linear_interpolation_test
>>> linked_list_test
>>> location_2d_test
>>> ltc2984_test
>>> matrix_test
>>> matrix_vector_test
>>> mcp2515_can_bit_timings_test
>>> median_test
>>> message_test
>>> moving_average_test
>>> node_test
>>> operator_test
>>> pair_test
>>> periodic_timer_test
>>> pid_test
>>> platform_gpio_test
FAIL: platform_gpio_test:66 : true == false
FAIL: platform_gpio_test:74 : true == false
FAIL: platform_gpio_test:79 : true == false
>>> point_set_2d_test
>>> polygon_2d_test
>>> prescaler_test
>>> protothread_test
>>> quaternion_test
>>> ramp_test
>>> range_test
>>> ray_2d_test
>>> register_test
>>> resumable_test
>>> s_curve_controller_test
>>> s_curve_generator_test
>>> saturated_test
>>> scheduler_test
>>> slave_test
>>> spi_bitbang_test
  0 -1694710967      0 (0us)
  1 1694716001   5034 (7us)
  2 -1694721015   5014 (7us)
  3 1694726030   5015 (7us)
  4 -1694730945   4915 (6us)
  5 1694735960   5015 (7us)
  6 -1694740875   4915 (6us)
  7 1694745890   5015 (7us)
  8 -1694750884   4994 (7us)
  9 1694755820   4936 (6us)
 10 -1694760814   4994 (7us)
 11 1694765750   4936 (6us)
 12 -1694770744   4994 (7us)
 13 1694775758   5014 (7us)
 14 -1694780673   4915 (6us)
 15 1694785688   5015 (7us)
 16 -1694790603   4915 (6us)

  0 -1694710967      0 (0us)

  0 -1694890485      0 (0us)
  1 1694890656    171 (1us)
  2 -1694900795  10139 (16us)
  3 1694910821  10026 (14us)
  4 -1694920843  10022 (14us)
  5 1694930868  10025 (14us)
  6 -1694940887  10019 (14us)
  7 1694950989  10102 (15us)
  8 -1694961012  10023 (14us)

  0 -1695029913      0 (0us)
  1 1695059993  30080 (0us)
  2 -1695099807  39814 (12us)

  0 -1695134603      0 (0us)
  1 1695144740  10137 (16us)
  2 -1695184619  39879 (12us)

>>> spi_device_test
>>> time_test
>>> timeout_test
>>> timestamp_test
>>> vector1_test
>>> vector2_test
>>> vector3_test
>>> vector4_test
>>> vector_test

Failed 4 of 5755 tests
FAIL!
```

---

Update:

> FAIL: angle_test:40 : -3.14159e+00 == 3.14159e+00

```c++
TEST_ASSERT_EQUALS_FLOAT(modm::Angle::perpendicular(-0.5 * M_PI, true),  1.0 * M_PI);
```


> FAIL: platform_gpio_test:66 : true == false
> FAIL: platform_gpio_test:74 : true == false
> FAIL: platform_gpio_test:79 : true == false

```c++
void
PlatformGpioTest::testEnabled()
{
	// checks if the handler in enable.cpp has been called
	// if this fails, then the function either wasn't found by the linker
	// or the section was garbage collected, or the startup script has a bug
	TEST_ASSERT_TRUE(RCC->APB2ENR & RCC_APB2ENR_SYSCFGEN);
	TEST_ASSERT_TRUE(SYSCFG->CMPCR & SYSCFG_CMPCR_CMP_PD);
	// check all ports are clock enabled and all GPIOs are reset
	TEST_ASSERT_TRUE((GPIOA->MODER & (3ul << 0)) == 0ul);  // line 66
	TEST_ASSERT_TRUE((GPIOA->OSPEEDR & (3ul << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->PUPDR & (3ul << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->OTYPER & (1ul << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->AFR[0] & (0xful << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->ODR & (1ul << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->BSRR & (0x10001ul << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->LCKR & (1ul << 0)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->MODER & (3ul << 2)) == 0ul);  // line 74
	TEST_ASSERT_TRUE((GPIOA->OSPEEDR & (3ul << 2)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->PUPDR & (3ul << 2)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->OTYPER & (1ul << 1)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->AFR[0] & (0xful << 4)) == 0ul);
	TEST_ASSERT_TRUE((GPIOA->ODR & (1ul << 1)) == 0ul); // line 79
```

chris-durand commented:
BlockAllocator is fixed with https://github.com/modm-io/modm/pull/529. The implementation assumed `MODM_ALIGN >= 4` which is not true on AVR.

chris-durand commented:
I am fixing AVR `time_test`. It dereferences a pointer to a value in flash without accessor::Flash. Furthermore, the date implementation has some integer overflows on AVR.

We should consider to move to llvm for AVR at some point because they have solved the problem with the separated flash address space in a beautiful way, no need for `accessor::Flash` anymore. With llvm the address space attribute is part of the type system. Thus, you can even run STL algorithms on flash arrays. See here: https://gcc.godbolt.org/z/cTjWYn

chris-durand commented:
Fix for Date classes on AVR: https://github.com/modm-io/modm/pull/530
Note to self: this Date stuff needs refactoring, let's fix the bugs first ...

rleh commented:
@salkinium Are you ok with merging this PR despite of 169dfab?

salkinium commented:
> @salkinium Are you ok with merging this PR despite of 169dfab?

Yes.

chris-durand commented:
Apart from the communication tests everything passes on AVR with https://github.com/modm-io/modm/pull/531 and the branch rebased to develop.

```
Unittests (Jan  6 2021, 23:29:30)
>>> LUDecomposition_test
>>> ad7280a_test
>>> angle_test
>>> arithmetic_traits_test
>>> atomic_queue_test
>>> bit_operation_test
>>> block_allocator_test
>>> bounded_deque_test
>>> bounded_queue_test
>>> bounded_stack_test
>>> button_group_test
>>> button_test
>>> can_bit_timings_test
>>> can_lawicel_formatter_test
>>> can_message_test
>>> circle_2d_test
>>> clock_test
>>> debounce_test
>>> doubly_linked_list_test
>>> drv832x_spi_test
>>> dynamic_array_test
>>> endianness_test
>>> fir_test
>>> flash_test
>>> i2c_test
>>> io_stream_test
>>> lagrange_interpolation_test
>>> line_2d_test
>>> line_segment_2d_test
>>> linear_interpolation_test
>>> linked_list_test
>>> location_2d_test
>>> ltc2984_test
>>> matrix_test
>>> matrix_vector_test
>>> mcp2515_can_bit_timings_test
>>> median_test
>>> moving_average_test
>>> operator_test
>>> pair_test
>>> periodic_timer_test
>>> pid_test
>>> point_set_2d_test
>>> polygon_2d_test
>>> prescaler_test
>>> protothread_test
>>> quaternion_test
>>> ramp_test
>>> range_test
>>> ray_2d_test
>>> register_test
>>> resumable_test
>>> s_curve_controller_test
>>> s_curve_generator_test
>>> saturated_test
>>> scheduler_test
>>> spi_device_test
>>> time_test
>>> timeout_test
>>> timestamp_test
>>> vector1_test
>>> vector2_test
>>> vector3_test
>>> vector4_test
>>> vector_test

Passed 3978 tests
OK!

```

chris-durand commented:
The xpcc AVR tests succeed whereas sab fails.
```
Unittests (Jan  7 2021, 00:11:55)
>>> can_connector_base_test
>>> can_connector_test
>>> dispatcher_test
>>> interface_test
>>> slave_test
FAIL: slave_test:118 : 0 == 1
FAIL: slave_test:140 : 0 == 2
FAIL: slave_test:148 : true == false
FAIL: slave_test:151 : 1 == 2
FAIL: slave_test:154 : 
[1, 163]
[205, 171]
FAIL: slave_test:166 : 0 == 3
FAIL: slave_test:190 : 0 == 4
FAIL: slave_test:191 : 0 == 39030
FAIL: slave_test:199 : true == false
FAIL: slave_test:202 : 1 == 4
FAIL: slave_test:204 : 
[1, 9, 0, 222]
[120, 86, 52, 18]

Failed 11 of 320 tests
FAIL!
```

chris-durand commented:
I don't intend to fix the SAB implementation because it's broken unrelated to this PR. The interface of `sab::Slave` requires the action array to be stored in flash and reads it using `modm::accessor::Flash`. The array content can't have constant initialization due to some undefined behavior `reinterpret_cast`s of member function pointers in the `SAB_ACTION` macro and thus, can't be declared as `PROGMEM`. Therefore, the code reads the data from flash although it actually is in RAM and fails.

rleh commented:
Thanks @chris-durand for the great help with the tests!

I suggest to merge the pull request soon and then see which further improvements or bug fixes come up in the future.

go2sh commented on tools/build_script_generator/common.py:
```diff
@@ -260,7 +260,8 @@ def common_compiler_flags(compiler, target):
         "-fstrict-enums",
-        "-std=c++17",
+        "-std=c++20",
+        "-Wno-volatile",  # volatile is deprecated in C++20 but lots of our external code uses it...
```
Note that not every thing volatile is deprecated, but nearly every peace of exiting embedded vendor code.

salkinium replied:
Yes, in theory a good idea, just not very practical: https://github.com/modm-io/modm/pull/521#issuecomment-748393084
